p2p/server.py

```diff
@@ -122,36 +143,121 @@ def __init__(self,
+        peer = self.peer_class(
+            remote=initiator_remote,
+            privkey=self.privkey,
+            reader=reader,
+            writer=writer,
+            aes_secret=aes_secret,
+            mac_secret=mac_secret,
+            egress_mac=egress_mac,
+            ingress_mac=ingress_mac,
+            chaindb=self.chaindb,
             network_id=self.network_id
         )
-        self.peer_pool.add_peer(eth_peer)
-
-
-def decode_authentication(ciphertext: bytes,
-                          privkey: datatypes.PrivateKey
-                          ) -> Tuple[datatypes.PublicKey, bytes, datatypes.PublicKey]:
-    """
-    Decrypts and decodes the ciphertext msg.
-    Returns the initiator's ephemeral pubkey, nonce, and pubkey.
-    """
-    if len(ciphertext) < ENCRYPTED_AUTH_MSG_LEN:
-        raise ValueError("Auth msg too short: {}".format(len(ciphertext)))
-    elif len(ciphertext) == ENCRYPTED_AUTH_MSG_LEN:
-        sig, initiator_pubkey, initiator_nonce, _ = decode_auth_plain(
-            ciphertext, privkey)
+
+        await self.apply_p2p_handshake(peer)
+
+    async def apply_p2p_handshake(self, peer: BasePeer) -> None:
+        finish_p2p_handshake = False
+        try:
+            # P2P Handshake.
+            await wait_with_token(
+                peer.do_p2p_handshake(),
+                token=peer.cancel_token,
+                timeout=HANDSHAKE_TIMEOUT,
+            )
+            finish_p2p_handshake = True
+        except (HandshakeFailure, asyncio.TimeoutError) as e:
+            self.logger.debug('Unable to finish P2P handshake: {}'.format(e))
+
+        # Run peer and add peer.
+        if finish_p2p_handshake:
+            self.peer_pool.start_peer(peer)
+
+
+def _test() -> None:
+    import argparse
+
+    from concurrent.futures import ProcessPoolExecutor
+
+    from eth_keys import keys
+    from eth_utils import (
+        decode_hex,
+    )
+
+    from evm.db.backends.memory import MemoryDB
+
+    from p2p import ecies
+    from p2p.exceptions import (
+        OperationCancelled,
+    )
+
+    from tests.p2p.dumb_peer import DumbPeer
+    from tests.p2p.integration_test_helpers import FakeAsyncChainDB
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-debug', action="store_true")
+    parser.add_argument('-server-address', type=str, required=True)
+    parser.add_argument('-bootstrap', '--list', type=str)
+    parser.add_argument('-privkey', type=str)
+
+    args = parser.parse_args()
+
+    logging.basicConfig(
+        level=logging.INFO, format='%(asctime)s %(levelname)s: %(message)s', datefmt='%H:%M:%S')
+    log_level = logging.INFO
+    if args.debug:
+        log_level = logging.DEBUG
+    logging.getLogger('p2p.server.Server').setLevel(log_level)
+
+    loop = asyncio.get_event_loop()
+    # Use a ProcessPoolExecutor as the default because the tasks we want to offload from the main
+    # thread are cpu intensive.
+    loop.set_default_executor(ProcessPoolExecutor())
```

Comment:
No need for a `ProcessPoolExecutor` here -- that is only used when we are doing a chain sync